authority/admin/collection.go

```diff
@@ -0,0 +1,180 @@
+
+func (p adminSlice) Len() int           { return len(p) }
+func (p adminSlice) Less(i, j int) bool { return p[i].uid < p[j].uid }
+func (p adminSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// Collection is a memory map of admins.
+type Collection struct {
+	byID                    *sync.Map
+	bySubProv               *sync.Map
+	byProv                  *sync.Map
+	sorted                  adminSlice
+	provisioners            *provisioner.Collection
+	superCount              int
+	superCountByProvisioner map[string]int
+}
+
+// NewCollection initializes a collection of provisioners. The given list of
+// audiences are the audiences used by the JWT provisioner.
+func NewCollection(provisioners *provisioner.Collection) *Collection {
+	return &Collection{
+		byID:                    new(sync.Map),
+		byProv:                  new(sync.Map),
+		bySubProv:               new(sync.Map),
+		superCountByProvisioner: map[string]int{},
+		provisioners:            provisioners,
+	}
+}
+
+// LoadByID a admin by the ID.
+func (c *Collection) LoadByID(id string) (*linkedca.Admin, bool) {
+	return loadAdmin(c.byID, id)
+}
+
+func subProvNameHash(sub, provName string) string {
+	subHash := sha1.Sum([]byte(sub))
+	provNameHash := sha1.Sum([]byte(provName))
+	_res := sha1.Sum(append(subHash[:], provNameHash[:]...))
+	return string(_res[:])
+}
+
+// LoadBySubProv a admin by the subject and provisioner name.
+func (c *Collection) LoadBySubProv(sub, provName string) (*linkedca.Admin, bool) {
+	return loadAdmin(c.bySubProv, subProvNameHash(sub, provName))
+}
+
+// LoadByProvisioner a admin by the subject and provisioner name.
+func (c *Collection) LoadByProvisioner(provName string) ([]*linkedca.Admin, bool) {
+	a, ok := c.byProv.Load(provName)
+	if !ok {
+		return nil, false
+	}
+	admins, ok := a.([]*linkedca.Admin)
+	if !ok {
+		return nil, false
+	}
+	return admins, true
+}
+
+// Store adds an admin to the collection and enforces the uniqueness of
+// admin IDs and amdin subject <-> provisioner name combos.
+func (c *Collection) Store(adm *linkedca.Admin) error {
+	p, ok := c.provisioners.Load(adm.ProvisionerId)
+	if !ok {
+		return fmt.Errorf("provisioner %s not found", adm.ProvisionerId)
+	}
+	// Store admin always in byID. ID must be unique.
+	if _, loaded := c.byID.LoadOrStore(adm.Id, adm); loaded {
+		return errors.New("cannot add multiple admins with the same id")
+	}
+
+	provName := p.GetName()
+	// Store admin always in bySubProv. Subject <-> ProvisionerName must be unique.
+	if _, loaded := c.bySubProv.LoadOrStore(subProvNameHash(adm.Subject, provName), adm); loaded {
+		c.byID.Delete(adm.Id)
+		return errors.New("cannot add multiple admins with the same subject and provisioner")
+	}
+
+	if admins, ok := c.LoadByProvisioner(provName); ok {
+		c.byProv.Store(provName, append(admins, adm))
+		c.superCountByProvisioner[provName]++
+	} else {
+		c.byProv.Store(provName, []*linkedca.Admin{adm})
+		c.superCountByProvisioner[provName] = 1
```

Comment:
Writing and reading from this map using different is not safe.

Reply:
what do you mean by using different? Do you want me to use a sync map?

Reply:
Ups I meant using different goroutines. I think we are now locking at a different level so it should be ok.